src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,347 @@
+    kj::Array<kj::byte> serializedException;
+  };
+  // The serialized JS value.
+  using SerializedResult = kj::OneOf<Returned, ThrewException>;
+
+  // Invokes the given function and serializes the result for RPC.
+  jsg::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    // `processAsyncResult` expects a promise, so even if we return synchronously we want to wrap
+    // the result of calling the JS function in a promise.
+    auto wrapValInPromise = [](jsg::Lock& js, v8::Local<v8::Value> value) {
+      auto resolver = jsg::check(v8::Promise::Resolver::New(js.v8Context()));
+      auto promise = resolver->GetPromise();
+
+      jsg::check(resolver->Resolve(js.v8Context(), value));
+      return js.toPromise(kj::mv(promise));
+    };
+
+    kj::Maybe<jsg::Promise<jsg::Value>> eventualResult;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      eventualResult = wrapValInPromise(js, invokeFn(js, fn, serializedArgs));
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      eventualResult = wrapValInPromise(js, exception.getHandle(js));
+    });
+
+    auto& result = KJ_REQUIRE_NONNULL(eventualResult);
+    return processAsyncResult(js, kj::mv(result)).then(js,
+        [caughtException](jsg::Lock& js, SerializedResult serializedResult) -> SerializedResult {
+          if (caughtException) {
+            KJ_IF_SOME(ser, serializedResult.tryGet<Returned>()) {
+              // Since we threw while executing the function, we still want to throw the
+              // serialized result even though serialization was successful.
+              return ThrewException { .serializedException = kj::mv(ser.serializedValue) };
+            }
+          }
+          return kj::mv(serializedResult);
+        });
+  }
+
+  // The JS method returned a promise and we need to wait for the promise to resolve or reject
+  // before serializing the result.
+  jsg::Promise<SerializedResult> processAsyncResult(
+      jsg::Lock& js,
+      jsg::Promise<jsg::Value> promise) {
+    return promise.then(js, [] (jsg::Lock& lock, jsg::Value value) -> SerializedResult {
+      auto jsVal = jsg::JsValue(value.getHandle(lock));
+      return Returned { .serializedValue = serializeV8(lock, jsVal) };
+    }).catch_(js, [](jsg::Lock& lock, jsg::Value exception) -> SerializedResult {
+      auto jsVal = jsg::JsValue(exception.getHandle(lock));
+      return lock.tryCatch([&] {
+        // First, we'll try serializing this value.
+        return ThrewException { .serializedException = serializeV8(lock, jsVal) };
+      }, [&](jsg::Value) {
+        // We failed to serialize the result, let's just throw an exception back to the client.
+        lock.throwException(kj::mv(exception));
+        KJ_UNREACHABLE;
+        // Compiler wants to see a return type...
+        return ThrewException {};
```

Comment:
Declare the return type on the lambda.

```suggestion
      }, [&](jsg::Value) -> ThrewException {
        // We failed to serialize the result, let's just throw an exception back to the client.
        lock.throwException(kj::mv(exception));
```

(Note the `KJ_UNREACHABLE` is also not necessary because `lock.throwException()` is already declared `[[noreturn]]`.)